client/web/src/open-in-editor/migrate-legacy-settings.ts

```diff
@@ -0,0 +1,119 @@
+import { Settings } from '@sourcegraph/shared/src/settings/settings'
+
+import { EditorReplacements, EditorSettings } from './editor-settings'
+
+interface LegacySettings {
+    basePath?: string
+    linuxBasePath?: string
+    macBasePath?: string
+    windowsBasePath?: string
+    replacements?: EditorReplacements
+    editorId?: string
+    customUrlPattern?: string
+    vscodeUseInsiders?: boolean
+    vscodeUseSSH?: boolean
+    vscodeRemoteHost?: string
+    jetbrainsForceApi?: 'builtInServer'
+}
+
+/**
+ * Leaves the original Settings object unchanged.
+ * Returns a shallow copy with the old settings removed and the new ones added.
+ */
+export function migrateLegacySettings(settings: Settings): Settings {
+    if (settings.openInEditor !== undefined) {
+        return settings
+    }
+
+    const oldEditorSettingsInNewFormat = getOldSettingsInNewFormat(settings)
+
+    const newSettings: Settings = {
+        ...settings,
+        openInEditor: oldEditorSettingsInNewFormat,
+    }
+    for (const key of Object.keys(newSettings)) {
+        if (
+            key.startsWith('openineditor.') ||
+            key.startsWith('vscode.open.') ||
+            key.startsWith('openInIntellij.') ||
+            key.startsWith('openInWebstorm.') ||
+            key.startsWith('openInAtom.')
+        ) {
+            delete newSettings[key]
+        }
+    }
+
+    return newSettings
+}
+
+function getOldSettingsInNewFormat(settings: Settings): EditorSettings {
+    const old = {
+        ...readAtomSettings(settings),
```

Comment:
I feel like creating one curried function for every field is a bit much, should we inline this instead?

e.g.

```suggestion
        ...readLegacySettings(settings, 'openInAtom'),
```